Fix OpenStack image v2 pagination handling

toabctl opened the pull request:
When listing all images with "list_images()", the pagination needs to
be handled correctly. The image V2 API [0] has a "next" link if more
images are available. So listing *all* images need to loop until there
is no more "next" link available.

[0] https://docs.openstack.org/api-ref/image/v2/#list-images

toabctl committed:
Fix OpenStack image v2 pagination handling

When listing all images with "list_images()", the pagination needs to
be handled correctly. The image V2 API [0] has a "next" link if more
images are available. So listing *all* images need to loop until there
is no more "next" link available.

[0] https://docs.openstack.org/api-ref/image/v2/#list-images

Fixes #1466

vdloo commented:
would be nice to see a test for this in `libcloud/test/compute/test_openstack.py`

toabctl commented:
> would be nice to see a test for this in `libcloud/test/compute/test_openstack.py`

@vdloo done. But I just extended the currently available test. I'm not sure how to add another test (and use a different mock) for `list_images()`. Either you are fine with what I did or you need to help a bit how to create a second `list_images()` fixture/method invokation.

vdloo commented:
cool, looks good to me :+1:

before:
```
In [5]: conn.list_images()
Out[5]: 
[<NodeImage: id=44864339-18a3-445f-bf91-1398e0ad4d32, name=image1000, driver=OpenStack  ...>,
 <NodeImage: id=e1ad28c3-8770-439a-9120-6261bdbc8ad6, name=image999, driver=OpenStack  ...>,
 <NodeImage: id=d6d02d2b-89c7-4c22-b03d-5990469e15d0, name=image998, driver=OpenStack  ...>,
 <NodeImage: id=8b6de47f-59be-4c61-86cf-c5366b9d65c3, name=image997, driver=OpenStack  ...>,
 <NodeImage: id=ccbd11f0-025b-4b9d-b9af-8f3d9729a5bb, name=image996, driver=OpenStack  ...>,
 <NodeImage: id=30cc7e45-ecaa-46de-a852-44674899f392, name=image995, driver=OpenStack  ...>,
 <NodeImage: id=962dbfe1-3cf6-4779-81a4-1ff053adb554, name=image994, driver=OpenStack  ...>,
 <NodeImage: id=c6d6331e-e91f-4e69-a773-4a35345124a9, name=image993, driver=OpenStack  ...>,
 <NodeImage: id=8bcd3aaf-03e9-4811-9761-3eea1c17796d, name=image992, driver=OpenStack  ...>,
 <NodeImage: id=b81b9757-590a-44d2-b60e-7661a1e8f8dc, name=image991, driver=OpenStack  ...>,
 <NodeImage: id=d98c6664-b3dc-44b3-9829-a26324e7d02b, name=image990, driver=OpenStack  ...>,
 <NodeImage: id=035406b9-7e84-4ca8-a427-eefe29fa45a0, name=image989, driver=OpenStack  ...>,
 <NodeImage: id=641d061c-9891-4211-81b2-02396979eb40, name=image988, driver=OpenStack  ...>,
 <NodeImage: id=58e43728-813e-4353-9504-84b1aa8bd91d, name=image987, driver=OpenStack  ...>,
 <NodeImage: id=a0c78610-6fea-4638-b95e-10cea6968853, name=image986, driver=OpenStack  ...>,
 <NodeImage: id=c5dc07a4-3c2f-4444-8c3e-b8cff9f0bc22, name=image985, driver=OpenStack  ...>,
 <NodeImage: id=e1880ebf-4940-4d65-b31e-fa0b4d66e939, name=image984, driver=OpenStack  ...>,
 <NodeImage: id=32686def-0a8c-47cf-9609-68740f27b475, name=image983, driver=OpenStack  ...>,
 <NodeImage: id=99c0aa14-5bd1-45bc-b5ae-ec024c3e74b4, name=image982, driver=OpenStack  ...>,
 <NodeImage: id=ad328d5d-7743-43fc-ac41-db1fa57edf8a, name=image981, driver=OpenStack  ...>,
 <NodeImage: id=4d1bc2bd-098d-49e2-8a05-f622c74e9332, name=image980, driver=OpenStack  ...>,
 <NodeImage: id=51428528-676a-4403-a013-1979111d090f, name=image979, driver=OpenStack  ...>,
 <NodeImage: id=d82e3e3a-1f40-4dca-9a48-a6bf37d07669, name=image978, driver=OpenStack  ...>,
 <NodeImage: id=6d7ce1c0-2200-49b9-8e6d-ed8db80378ba, name=image977, driver=OpenStack  ...>,
 <NodeImage: id=e419dc3f-6989-46b8-87c2-bf930722ccb9, name=image976, driver=OpenStack  ...>]

In [6]: len(conn.list_images())
Out[6]: 25
```

after:
```
In [2]: conn.list_images()
Out[2]:
[<NodeImage: id=44864339-18a3-445f-bf91-1398e0ad4d32, name=image1000, driver=OpenStack  ...>,
 <NodeImage: id=e1ad28c3-8770-439a-9120-6261bdbc8ad6, name=image999, driver=OpenStack  ...>,
 <NodeImage: id=d6d02d2b-89c7-4c22-b03d-5990469e15d0, name=image998, driver=OpenStack  ...>,
 <NodeImage: id=8b6de47f-59be-4c61-86cf-c5366b9d65c3, name=image997, driver=OpenStack  ...>,
 <NodeImage: id=ccbd11f0-025b-4b9d-b9af-8f3d9729a5bb, name=image996, driver=OpenStack  ...>,
 <NodeImage: id=30cc7e45-ecaa-46de-a852-44674899f392, name=image995, driver=OpenStack  ...>,
 <NodeImage: id=962dbfe1-3cf6-4779-81a4-1ff053adb554, name=image994, driver=OpenStack  ...>,
 <NodeImage: id=c6d6331e-e91f-4e69-a773-4a35345124a9, name=image993, driver=OpenStack  ...>,
 <NodeImage: id=8bcd3aaf-03e9-4811-9761-3eea1c17796d, name=image992, driver=OpenStack  ...>,
 <NodeImage: id=b81b9757-590a-44d2-b60e-7661a1e8f8dc, name=image991, driver=OpenStack  ...>,
 <NodeImage: id=d98c6664-b3dc-44b3-9829-a26324e7d02b, name=image990, driver=OpenStack  ...>,
 <NodeImage: id=035406b9-7e84-4ca8-a427-eefe29fa45a0, name=image989, driver=OpenStack  ...>,
 <NodeImage: id=641d061c-9891-4211-81b2-02396979eb40, name=image988, driver=OpenStack  ...>,
 <NodeImage: id=58e43728-813e-4353-9504-84b1aa8bd91d, name=image987, driver=OpenStack  ...>,
 <NodeImage: id=a0c78610-6fea-4638-b95e-10cea6968853, name=image986, driver=OpenStack  ...>,
 <NodeImage: id=c5dc07a4-3c2f-4444-8c3e-b8cff9f0bc22, name=image985, driver=OpenStack  ...>,
 <NodeImage: id=e1880ebf-4940-4d65-b31e-fa0b4d66e939, name=image984, driver=OpenStack  ...>,
 <NodeImage: id=32686def-0a8c-47cf-9609-68740f27b475, name=image983, driver=OpenStack  ...>,
 <NodeImage: id=99c0aa14-5bd1-45bc-b5ae-ec024c3e74b4, name=image982, driver=OpenStack  ...>,
 <NodeImage: id=ad328d5d-7743-43fc-ac41-db1fa57edf8a, name=image981, driver=OpenStack  ...>,
 <NodeImage: id=4d1bc2bd-098d-49e2-8a05-f622c74e9332, name=image980, driver=OpenStack  ...>,
 <NodeImage: id=51428528-676a-4403-a013-1979111d090f, name=image979, driver=OpenStack  ...>,
 <NodeImage: id=d82e3e3a-1f40-4dca-9a48-a6bf37d07669, name=image978, driver=OpenStack  ...>,
 <NodeImage: id=6d7ce1c0-2200-49b9-8e6d-ed8db80378ba, name=image977, driver=OpenStack  ...>,
 <NodeImage: id=e419dc3f-6989-46b8-87c2-bf930722ccb9, name=image976, driver=OpenStack  ...>,
 <NodeImage: id=2a37409b-bdee-41ff-b564-4ca4e795a7d4, name=image975, driver=OpenStack  ...>,
...
In [3]: len(conn.list_images())
Out[3]: 1060
```

vdloo commented:
@Kami what's the process for merging these days? Just merge to master and someone (you) updates [CHANGES.rst](https://github.com/apache/libcloud/commits/trunk/CHANGES.rst)? before merging on github was allowed commits had to be signed as well iirc, is that no longer the case?

Kami commented:
@vdloo Yeah, merging into master directly via Github web is fine.

And either the contributor or person who is merging the PR should update CHANGES.rst file.

Kami commented on libcloud/compute/drivers/openstack.py:
```diff
@@ -2862,10 +2862,14 @@ def list_images(self, location=None, ex_only_active=True):
-        for image in response.object['images']:
-            images.append(self._to_image(image))
+        path = '/v2/images'
+        while path:
```
I'm a bit worried in this loop potentially never terminating if API returns an invalid response (aka incorrectly includes ``next`` attribute or similar).

It looks like we already "_paginated_request" utility method which should handle that scenario (https://github.com/apache/libcloud/blob/6da1d6be5fcec4d25ff71b6cbfbf78f409283888/libcloud/compute/drivers/openstack.py#L176).

So probably we should utilize it here as well.

Kami replied:
@vdloo Do you happen to have any OpenStack cluster you could use to test the "_paginated_request" approach?

vdloo replied:
yes I do, but _paginated_request works slightly different though (would expect images_links, see my comment [here](https://github.com/apache/libcloud/issues/1466) and [here](https://github.com/apache/libcloud/pull/1263))

a version like
```python
    def list_images(self, location=None, ex_only_active=True):
        """
        Lists all active images using the V2 Glance API

        @inherits: :class:`NodeDriver.list_images`

        :param location: Which data center to list the images in. If
                               empty, undefined behavior will be selected.
                               (optional)
        :type location: :class:`.NodeLocation`

        :param ex_only_active: True if list only active (optional)
        :type ex_only_active: ``bool``
        """
        if location is not None:
            raise NotImplementedError(
                "location in list_images is not implemented "
                "in the OpenStack_2_NodeDriver")
        if not ex_only_active:
            raise NotImplementedError(
                "ex_only_active in list_images is not implemented "
                "in the OpenStack_2_NodeDriver")

        response = self._paginated_request(
            '/v2/images', 'images', self.network_connection)
        return [self._to_image(port) for image in response['images']]
```
would result in:

```python
----> 1 conn.list_images()

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/compute/drivers/openstack.py in list_images(self, location, ex_only_active)
   2846 
   2847         response = self._paginated_request(
-> 2848             '/v2/images', 'images', self.network_connection)
   2849         return [self._to_image(port) for image in response['images']]
   2850 

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/compute/drivers/openstack.py in _paginated_request(url, obj, connection, params)
    198         loop_count = 0
    199         while True:
--> 200             data = connection.request(url, params=params)
    201             values = data.object.get(obj, list())
    202             objects.extend(values)

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/common/openstack.py in request(self, action, params, data, headers, method, raw) 
    223                                                             method=method,
    224                                                             headers=headers,
--> 225                                                             raw=raw)
    226 
    227     def _get_auth_url(self):

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/common/base.py in request(self, action, params, data, headers, method, raw, stream)
    636
    637         try:
--> 638             response = responseCls(**kwargs)
    639         finally:
    640             # Always reset the context after the request has completed

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/compute/drivers/openstack.py in __init__(self, *args, **kwargs)
   1122         # NodeDriver -> Connection -> Response
   1123         self.node_driver = OpenStack_1_1_NodeDriver
-> 1124         super(OpenStack_1_1_Response, self).__init__(*args, **kwargs)
   1125
   1126 

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/common/base.py in __init__(self, response, connection)
    149         if not self.success():
    150             raise exception_from_message(code=self.status,
--> 151                                          message=self.parse_error(),
    152                                          headers=self.headers)
    153 

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/common/openstack.py in parse_error(self)
    392             key_pair_name = context.get('key_pair_name', None)
    393 
--> 394             if len(values) > 0 and 'code' in values[0] and \
    395                     values[0]['code'] == 404 and key_pair_name:
    396                 raise KeyPairDoesNotExistError(name=key_pair_name,

TypeError: argument of type 'int' is not iterable
```
but perhaps modifying _paginated_request to also take into account this 'next' attribute would be a better solution than having this loop in the list_images method indeed

vdloo replied:
@toabctl do you want to give implementing this in terms of _paginated_request a try and also add a line to CHANGES.rst for this change?

toabctl replied:
The method looks pretty special. I can try to extract the endless loop detection and reuse that then.

Kami committed:
Add info about the new release to the doap file.

Kami committed:
Merge branch 'fix-openstack-imagev2-pagination' of https://github.com/toabctl/libcloud into toabctl-fix-openstack-imagev2-pagination

Kami committed:
Add new "_paginated_request_next" method and update list_images() method
to utilize it.

This method includes infinite loop protection and makes sure driver
doesn't hang / end up in an infinite loop if the API returns an
invalid response or similar.

Kami commented:
I pushed a small change so ``list_images()`` now utilizes a utility method which includes infinite loop protection - 7c503a3e695a7af05a8e922b6f27043732319878.

I still need to add tests, then it should be ready to be merged.

Kami committed:
Add a test case which verifies that the infinite loop protection in
"_paginated_request_next" works correctly if the "next" attribute
matches current url.

Kami committed:
Make code easier to read by moving a conditional in a separatem
top-level if statement.

Kami commented:
@vdloo Let me know if you will have a chance to test the changes above^.

I think bailing out if there are more than 1000 pages is reasonable enough (that means 1000 * page size elements total which is plenty).

vdloo commented:
yup, works fine:
```
In [2]: len(conn.list_images())                                                                            
Out[2]: 1113
```

and with `PAGINATION_LIMIT = 10`:
```
In [3]: conn.list_images()                                                                                 
---------------------------------------------------------------------------
OpenStackException                        Traceback (most recent call last)
<ipython-input-3-6e1a5f4cb5e9> in <module>
----> 1 conn.list_images()

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/compute/drivers/openstack.py in list_images(self, location, ex_only_active)
   2881             path='/v2/images',
   2882             request_method=self.image_connection.request,
-> 2883             response_key='images')
   2884 
   2885         images = []

~/.virtualenvs/libcloudtest/src/apache-libcloud/libcloud/compute/drivers/openstack.py in _paginated_request_next(self, path, request_method, response_key)
   2850                     'looping next target for pagination!' % (
   2851                         path, PAGINATION_LIMIT
-> 2852                     ), None
   2853                 )
   2854 

OpenStackException: 'Pagination limit reached for /v2/images?marker=1151542d-017c-47dd-94d2-f0550bbbc626, the limit is 10. This might indicate that your API is returning a looping next target for pagination!'
```

toabctl commented:
@kami @vdloo thanks for helping with that!

Kami commented:
Thanks for the contribution and thanks to @vdloo for a review and testing the changes.
